repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java

```diff
@@ -2628,12 +2629,62 @@ private void addCategoriesToTermEntity(AttributeMutationContext ctx, List<Object
         RequestContext.get().endMetricRecord(metricRecorder);
     }
 
-    private void addMeaningsToEntity(AttributeMutationContext ctx, List<Object> createdElements, List<AtlasEdge> deletedElements) {
-        addMeaningsToEntity(ctx, createdElements, deletedElements, false);
+
+    private void addMeaningsToEntityRelations(AttributeMutationContext ctx, List<Object> createdElements, List<AtlasEdge> deletedElements, boolean isAppend) {
+        MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("addMeaningsToEntityRelations");
+        // handle __terms attribute of entity
+        List<AtlasVertex> assignedEntitiesVertices ;
+        List<String> currentMeaningsQNames ;
+        String newMeaningName = ctx.referringVertex.getProperty(NAME, String.class);
+        String newMeaningPropertyKey = ctx.referringVertex.getProperty(QUALIFIED_NAME, String.class);
+
+        if (!createdElements.isEmpty()) {
+            assignedEntitiesVertices = createdElements.stream()
+                    .map(x -> ((AtlasEdge) x).getInVertex())
+                    .filter(x -> ACTIVE.name().equals(x.getProperty(STATE_PROPERTY_KEY, String.class)))
+                    .collect(Collectors.toList());
+
+            for (AtlasVertex relationVertex : assignedEntitiesVertices) {
+                currentMeaningsQNames = relationVertex.getMultiValuedProperty(MEANINGS_PROPERTY_KEY, String.class);
+
+                if (!isAppend) {
+                    relationVertex.removeProperty(MEANINGS_PROPERTY_KEY);
+                    relationVertex.removeProperty(MEANINGS_TEXT_PROPERTY_KEY);
+                }
+
+                if (!currentMeaningsQNames.contains(newMeaningPropertyKey)) {
+                    AtlasGraphUtilsV2.addEncodedProperty(relationVertex, MEANINGS_PROPERTY_KEY, newMeaningPropertyKey);
+                    AtlasGraphUtilsV2.setEncodedProperty(relationVertex, MEANINGS_TEXT_PROPERTY_KEY, newMeaningName);
+                    AtlasGraphUtilsV2.addListProperty(relationVertex, MEANING_NAMES_PROPERTY_KEY, newMeaningName, true);
+                }
```

Comment:
Imagine
1. isAppend is false -> removeProperty will be called
2. Sent request to add asset_0 (update term_0)
3. asset_0 already have a couple of items MEANINGS_PROPERTY_KEY = [term_0_qn, term_1_qn]

With current logic, when relationVertex == asset_0
- newMeaningPropertyKey = term_0_qn
- currentMeaningsQNames = [term_0_qn, term_1_qn]
- Since removeProperty gets called, properties are empty now
- !currentMeaningsQNames.contains(newMeaningPropertyKey) -> false
- skips all 3 addEncodedProperty, setEncodedProperty, addListProperty calls
- Final state asset_0.MEANINGS_PROPERTY_KEY = []

Reply:
IMO, following should be put inside `if (!currentMeaningsQNames.contains(newMeaningPropertyKey)) {`
```if (!isAppend) {
                    relationVertex.removeProperty(MEANINGS_PROPERTY_KEY);
                    relationVertex.removeProperty(MEANINGS_TEXT_PROPERTY_KEY);
                }